### ISSUE TITLE
[Docs] Multitask GP & classification with missing values

### ISSUE THREAD

**Issue by choltz95** [cornellius-gp/gpytorch#851]
Hi, I am new to this framework & GP in general - I am not sure if this is the right place to ask this kind of question.

I am trying to perform online classification on a variable-length multivariate timeseries dataset with missing values. I am following the method of this paper:[ https://arxiv.org/abs/1706.04152]( https://arxiv.org/abs/1706.04152) - i.e. my framework is a multitask gp -> pytorch model

I have followed the documentation closely & defined my interpolation model this way:
```
class MTGPInterpolationModel(gpytorch.models.ExactGP):
    def __init__(self, input_dim, train_x, train_y, likelihood):
        super(MTGPInterpolationModel, self).__init__(train_x, train_y, likelihood)
...
```
Then, in a parent model, I want to do inference with something like

`output = self.predictor_model(self.interpolation_model(vital_features))`

for a single multivariate sequence described by vital_features and backprop a classification loss through the entire framework. My questions are:

1. Is something like this possible? I want to backprop through the interpolation module, but still need to define a likelihood loss?
2. Do I have to instantiate a new interpolation model/add to the existing model's data for each iteration/observation of my training procedure (I guess this might be the purpose of get_fantasy_model, but it seem to have issues with the multitask gp [/issues/800])?
4. I could not find any documentation on dealing with irregularly sampled observations for multitask GP. What is the recommended approach? what should the representation be for train_x (right now a multivariate temporal sequence with missing values)

Sorry for the basic questions. I really appreciate your help!

**Comment by jacobrgardner** [comment]
Hi @choltz95,

This should be relatively straight forward, as the `MultivariateNormal` distribution returned by the GP in eval mode is differentiable in every way: you can backprop through the mean, variance, samples drawn using `rsample`, or any pytorch function of these quantities.

The only minor consideration is that if you want to backprop through the predictive distribution in GPyTorch, you need to let us know not to detach the caches we compute at test time so that you’ll get gradients all the way back to the GP hyperparameters. There is a `gpytorch.settings.detach_test_caches` context you can set to `False` to accomplish this. Additionally, you’ll want to clear the test time caches any time the GP hyperparameters are changed (for example, each iteration of training). To do this, you can just change your model back to train mode and then to eval mode again, e.g.:
```python
model.train()
model.eval()
```
Just let me know if this isn’t sufficient detail and I can put together a quick example notebook of this early next week; see also my more detailed response here: https://github.com/cornellius-gp/gpytorch/issues/772 where I give a working example snippet for an `NN -> GP -> NN` model (just get rid of the first NN.

**Comment by choltz95** [comment]
Hi @jacobrgardner. Thanks a lot for your response. I really appreciate it. It's very clear & the notebook you provided is almost identical to the framework I am adopting. 

Do you happen have any advice for dealing with sparsely distributed missing values in the training data? In particular I am dealing with multivariate timeseries that could be composed something like:
s_1 = 1, 2, 3, np.nan, np.nan, 6, 7, ...
s_2 = 1, np.nan, 3, 4, 5, 6, ...

I understand that given learned parameters I can sample covariates from the model at regular time intervals, and I was hoping to apply a multitask gp to do this, but composing the input is unclear to me. Is there a way to learn from this irregular data without removing timesteps containing nans at all, or is this the only option?

Thanks again for your help. Your library is really great.

**Comment by choltz95** [comment]
After reading the documentation more carefully: 

`Kernel supporting Kronecker style multitask Gaussian processes (where every data point is evaluated at every task)`

So is it correct to say that the functionality for multitask kernel to deal with irregularly distributed missing values is not yet supported? Or is there another option?

**Comment by Balandat** [comment]
You can use a hadarmard style GP for non-regular X: https://github.com/cornellius-gp/gpytorch/blob/master/examples/03_Multitask_GP_Regression/Hadamard_Multitask_GP_Regression.ipynb

**Comment by choltz95** [comment]
Thanks a lot. It works great after I sample an equal number of observations for each task - I guess this is the only way to approach an inconsistent # of observations for each task.

**Comment by choltz95** [comment]
Thanks a lot!
